honeybadgermpc/field.py

```diff
@@ -239,20 +239,42 @@ def __rdiv__(self, other):
 
     def sqrt(self):
         """Square root.
-
         No attempt is made the to return the positive square root.
-
-        Computing square roots is only possible when the modulus
-        is a Blum prime (congruent to 3 mod 4).
         """
-        assert self.modulus % 4 == 3, "Cannot compute square " \
-            "root of %s with modulus %s" % (self, self.modulus)
-
-        # Because we assert that the modulus is a Blum prime
-        # (congruent to 3 mod 4), there will be no reminder in the
-        # division below.
-        root = pow(self.value, (self.modulus+1)//4, self.modulus)
-        return GFElement(root, self.field)
+        assert self.modulus % 2 == 1
+
+        if self.modulus % 4 == 3:
+            # The case that the modulus is a Blum prime
+            # (congruent to 3 mod 4), there will be no reminder in the
+            # division below.
+            root = pow(self.value, (self.modulus+1)//4)
+            return GFElement(root, self.field)
+        else:
+            # The case that self.modulus % 4 == 1
+            # Cipolla’s Algorithm
+            # http://people.math.gatech.edu/~mbaker/pdf/cipolla2011.pdf
+            t = u = 0
+            for i in range(1, self.modulus):
+                u = i*i - self
+                if pow(u, (self.modulus-1)//2) == self.modulus - 1:
+                    t = i
+                    break
+
+            def cipolla_mult(a, b, w):
+                return ((a[0]*b[0] + a[1]*b[1]*w), (a[0]*b[1] + a[1]*b[0]))
+
+            exp = (self.modulus+1)//2
+            exp = [int(x) for x in bin(exp)[2:]]
```

Comment:
You don't need to necessarily convert this to an int. You can do `exp_str = bin(exp)[2:]` and the check for `exp[i] == "0"` later. I would also request you to rename this variable to something else since you are using the same name just before this line.